### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by jackzampolin**
This refactor is starting with adding an extensive testing harness.

- [ ] Test harness spins up current signer
- [ ] Refactor signer command to cobra and rename to horcrux
- [ ] Add config mgmt (+viper) for the horcrux command
- [ ] Add state viewing and management to horcrux
- [ ] Refactor `internal/signer` to `signer` and potentially other packages

### CODE REVIEW

File: cmd/horcrux/cmd/config.go [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+		Use:     "init [chain-id] [chain-nodes]",
+		Aliases: []string{"i"},
+		Short:   "initalize configuration file and home directory if one doesn't already exist",
+		Long: "initalize configuration file, use flags for cosigner configuration.\n\n" +
+			"[chain-id] is the chain id of the chain to validate\n" +
+			"[chain-nodes] is a comma seperated array of chain node addresses i.e.\n" +
+			"tcp://chain-node-1:1234,tcp://chain-node-2:1234",
+		Args: cobra.RangeArgs(1, 2),
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			cid := args[0]
```
Comment (levicook) [inline cmd/horcrux/cmd/config.go:42]: FYI; This feels like it could panic with `runtime error: index out of range [0]` if a user runs it w/o providing an argument. (Maybe there's upstream protection on this tho)

File: testing/setup_test.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"github.com/testcontainers/testcontainers-go"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	chainid = "horcux"
```
Comment (levicook) [inline testing/setup_test.go:20]: Is `horcux` missing an `r`??

File: testing/setup_test.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	provider, err := testcontainers.NewDockerProvider()
+	require.NoError(t, err)
+
+	net, err := provider.CreateNetwork(ctx, testcontainers.NetworkRequest{Name: netid, Internal: false})
+	require.NoError(t, err)
+
+	nodes := MakeTestNodes(4, home, chainid, simdChain, provider)
+
+	cont := startValidatorContainers(t, provider, net, nodes)
+	require.NoError(t, err)
```
Comment (levicook) [inline testing/setup_test.go:45]: This line can be deleted; appears to be double checking what line 40 did above

File: testing/setup_test.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	})
+
+	for i, n := range nodes {
+		str, err := cont[i].PortEndpoint(ctx, "26657", "http")
+		require.NoError(t, err)
+		fmt.Printf("%s available at %s\n", n.Name(), str)
+		require.NoError(t, n.NewClient(str))
+	}
+
+	time.Sleep(10 * time.Second)
```
Comment (levicook) [inline testing/setup_test.go:62]: Why sleep here? Are you waiting for containers to start? I'm still learning the testcontainers package (which looks amazing) and think that their wait package might be helpful: github.com/testcontainers/testcontainers-go/wait

File: cmd/horcrux/cmd/config.go (hunk 1 above)
Comment (jackzampolin) [reply cmd/horcrux/cmd/config.go:42]: The range args function will prevent this

File: Makefile [hunk 5]
```diff
@@ -29,3 +22,9 @@ clean:
 	rm -rf build
 
 .PHONY: all lint test race msan tools clean build
+
+build-simd-docker:
+	docker build -t jackzampolin/simd:v0.42.3 -f ./docker/simd/Dockerfile ./docker/simd/
```
Comment (jackzampolin) [inline Makefile:27]: TODO: make version a variable at the top of makefile and keep up to date. Alternately pull in data from go.mod and make that the authoritative SDK version to run against:  `go list -m -u -f '{{.Version}}' github.com/cosmos/cosmos-sdk`. If taking second option please make sure that the docker files get built with the right arg and it gets used in the tests correctly

File: README.md [hunk 6]
```diff
@@ -1,8 +1,17 @@
-# Tendermint Validator
+# Horcrux
```
Comment (jackzampolin) [inline README.md:1]: This whole file needs to be rewitten. That is out of scope for this PR however.

File: cmd/horcrux/cmd/config.go [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+			"[chain-id] is the chain id of the chain to validate\n" +
+			"[chain-nodes] is a comma seperated array of chain node addresses i.e.\n" +
+			"tcp://chain-node-1:1234,tcp://chain-node-2:1234",
+		Args: cobra.RangeArgs(1, 2),
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			return nil
+		},
+	}
+	return cmd
+}
```
Comment (jackzampolin) [inline cmd/horcrux/cmd/config.go:46]: Remove these lines.

File: cmd/horcrux/cmd/config.go [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+				homeDir = cfgFile
+			} else {
+				home, _ := homedir.Dir()
+				homeDir = path.Join(home, ".horcrux")
+			}
+
+			//TODO double check that omitting directory check does not introduce unexpected behavior
+			//if _, err := os.Stat(homeDir); !os.IsNotExist(err) {
+			//	return fmt.Errorf("%s is not empty, check for existing configuration and clear path before trying again", homeDir)
+			//}
```
Comment (jackzampolin) [inline cmd/horcrux/cmd/config.go:79]: Lets re-add this? I think we handle this in root.go

File: cmd/horcrux/cmd/cosigner.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+import (
+	"fmt"
+	"log"
+	"net"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/jackzampolin/horcrux/internal/signer"
```
Comment (jackzampolin) [inline cmd/horcrux/cmd/cosigner.go:12]: NIT: we should move the directory from `internal/signer` to `signer` and replace the import path throughout the code.

File: cmd/horcrux/cmd/cosigner.go [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+	"github.com/spf13/cobra"
+	tmlog "github.com/tendermint/tendermint/libs/log"
+	tmOS "github.com/tendermint/tendermint/libs/os"
+	tmService "github.com/tendermint/tendermint/libs/service"
+	"github.com/tendermint/tendermint/types"
+)
+
+func init() {
+	cosignerCmd.AddCommand(signer.CreateCosignerSharesCmd())
```
Comment (jackzampolin) [inline cmd/horcrux/cmd/cosigner.go:21]: What is this? Prob remove?

File: cmd/horcrux/cmd/key2shares.go [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+	"github.com/spf13/cobra"
+	"github.com/tendermint/tendermint/crypto/ed25519"
+	tmjson "github.com/tendermint/tendermint/libs/json"
+	tmOS "github.com/tendermint/tendermint/libs/os"
+	"github.com/tendermint/tendermint/privval"
+	tsed25519 "gitlab.com/polychainlabs/threshold-ed25519/pkg"
+)
+
+// key2sharesCmd represents the key2shares command
+var key2sharesCmd = &cobra.Command{
```
Comment (jackzampolin) [inline cmd/horcrux/cmd/key2shares.go:36]: Currently this command will dump the key shares in the current directory. Maybe we want to dump them in the `~/.horcux/shares` dir or something? Current behavior is OK but maybe there is something better?

File: cmd/horcrux/cmd/key2shares.go [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+			}
+			fmt.Printf("Created Share %d\n", shareID)
+		}
+		fmt.Println("key2shares called")
+		return nil
+	},
+}
+
+// KeyToShares takes a priv key and some args and returns the share objects
+func KeyToShares(threshold, total int64, pvKey privval.FilePVKey) []signer.CosignerKey {
```
Comment (jackzampolin) [inline cmd/horcrux/cmd/key2shares.go:122]: Use this function in the above command, maybe in conjunction with the rethink of where the share files get put.

File: docker/simd/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,28 @@
+ARG VERSION=v0.42.3
```
Comment (jackzampolin) [inline docker/simd/Dockerfile:1]: Makefile should pass this arg from the go.mod and the tests should check the sdk version in go mod and build if the container doesn't exist.

File: docs/setup.md [hunk 14]
```diff
@@ -0,0 +1,111 @@
+## Pre-requisites
```
Comment (jackzampolin) [inline docs/setup.md:1]: This whole file needs to be rewitten.

File: internal/signer/CosignerKeyShares.go [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+
+	"github.com/spf13/cobra"
+	tmjson "github.com/tendermint/tendermint/libs/json"
+	"github.com/tendermint/tendermint/libs/os"
+	"github.com/tendermint/tendermint/privval"
+	tsed25519 "gitlab.com/polychainlabs/threshold-ed25519/pkg"
+)
+
+// CreateCosignerSharesCmd is a cobra command for creating cosigner shares from a priv validator
+func CreateCosignerSharesCmd() *cobra.Command {
```
Comment (jackzampolin) [inline internal/signer/CosignerKeyShares.go:19]: Actually we should use this cobra command instead!

File: scripts/cosigner1.toml [hunk 16]
```diff
@@ -14,7 +14,4 @@ id = 3
 remote_address = "tcp://0.0.0.0:2224"
```
Comment (jackzampolin) [inline scripts/cosigner1.toml:14]: Remove this file

File: scripts/cosigner2.toml [hunk 17]
```diff
@@ -12,3 +12,6 @@ remote_address = "tcp://0.0.0.0:2222"
 [[cosigner]]
```
Comment (jackzampolin) [inline scripts/cosigner2.toml:12]: remove this file

File: testing/node_test.go [hunk 18]
```diff
@@ -0,0 +1,491 @@
...
+	Repository string
+	Version    string
+	Bin        string
+	Ports      map[docker.Port]struct{}
+}
+
+// ChainType instance for simd
+var simdChain = &ChainType{
+	Repository: "jackzampolin/simd",
+	Version:    "v0.42.3",
```
Comment (jackzampolin) [inline testing/node_test.go:56]: Pull this version in dynamically from go.mod https://pkg.go.dev/golang.org/x/mod/modfile

File: cmd/horcrux/cmd/config.go (hunk 8 above)
Comment (jtieri) [reply cmd/horcrux/cmd/config.go]: I omitted these lines while fighting the permissions issues I was running into while accessing files created by Docker containers from Go. Now that containers are run from the user, instead of root, the fix for this is to not create the config directories when the signer structs are created and to let cmd/config handle it again
